docs/content/sips/xxx-spin-plugins.md

```diff
@@ -0,0 +1,200 @@
+
+These `types` are then grouped up into 2 categories called User Plugins and Internal Plugins depending on how they are invoked. Internal Plugins are called implicitly by application configuration while the User plugins are explicitly invoked by the user.
+
+Based on the type of plugin, Spin knows the execution path that needs to be followed before invoking the plugin (i.e) the environment variables that it sets and other things like placing files in appropriate directories. In the case of Internal plugins, the would also define how the data must be returned if any is required. 
+
+  - **Spin Dependant User Plugins** 
+These plugins are explicitly called by the user. An example of this would the `deploy` type of plugins (i.e) A potential plugin that deploys to application to bindle would be called using
+
+```
+spin deploy-bindle
+```
+
+where `deploy` would be the type of the plugin and spin would know the prerequisites that it must provide before invoking the plugin.
+
+  - **Spin Dependant Internal Plugins**
+These plugins are not explicitly called by the user but are implied implicitly based on the contents of `spin.toml`. (e.g) A trigger plugin based on time would be defined in the `spin.toml` as 
+
+```rust
+trigger = {type = "spin-trigger-time"}
+```
+
+In this case, Spin figures out that it needs to use an Internal Plugin to satisfy the requirements of the application. `trigger` here is an example of Internal Plugin `type`. 
+
+## Proposed implementation of `spin plugin`
+
+### The `install` subcommand
+
+When the `spin plugin install $plugin` command is invoked, spin uses a centralized repository to obtain information about the requested plugin.
+
+Once the information such as the source and version of the plugin is obtained, spin proceeds to download and place the binary in a spin managed directory along with the license.  
+
+#### Installing from other sources
+
+Optional arguments can be created to allow installation from a local directory or form a custom git repository.
+
+```
+# installing from local source
+spin plugin install -f <path_to_tar> <plugin_name>
+# installing from custom Git repo
+spin plugin install --custom <url_to_plugin_metadata>
+```
+
+#### Update of plugins
+
+The plugins can be updated using the `--update` flag on the `install` subcommand. The `--update` argument will update a package to the newest version if it already exists or else will install the newest version.
+
+**Dealing with Breaking changes**
+
+This is a section that still needs discussion. One of the solutions is for the plugin manifest to list the versions of spin that it is compatible with and use that to make an informed update of plugins avoiding updates to incompatible versions.
```

Comment:
I worry about synchronicity issues here. When a new version of spin is released, then all plugins will need to update their compat lists. This means the plugin owner will have to stay fairly involved in the plugins repository. I wonder what other package managers do. I know SPEC files allow for requires syntax but it doesnt always have to give version and sometimes it can be an equality operator (`requires: python >= 1.3`): http://ftp.rpm.org/api/4.4.2.2/dependencies.html

Reply:
The `semver` crate supports version constraints: https://docs.rs/semver/1.0.13/semver/struct.VersionReq.html